resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolverContext.java

```diff
@@ -144,15 +144,96 @@ public void operationComplete(Future<T> future) throws Exception {
     }
 
     private void internalResolve(Promise<T> promise) {
-        InetSocketAddress nameServerAddrToTry = nameServerAddrs.next();
+        DnsServerAddressStream nsAddrs = getNameServers(hostname);
+
         for (DnsRecordType type: parent.resolveRecordTypes()) {
-            if (!query(hostname, type, nameServerAddrToTry, promise)) {
+            if (!query(hostname, type, nsAddrs, promise)) {
                 return;
             }
         }
     }
 
-    private void query(InetSocketAddress nameServerAddr, final DnsQuestion question, final Promise<T> promise) {
+    /**
+     * Add an authoritative nameserver to the cache if its not a root server.
+     */
+    private void addNameServerToCache(
+            AuthoritativeNameServerList.AuthoritativeNameServer name, InetAddress resolved, long ttl) {
+        if (!name.isRootServer()) {
+            // Cache NS record if not for a root server as we should never cache for root servers.
+            parent.authoritativeDnsServerCache().cache(name.domainName(),
+                    additionals, resolved, ttl, parent.ch.eventLoop());
+        }
+    }
+
+    /**
+     * Returns the {@link DnsServerAddressStream} that was cached for the given hostname or {@code null} if non
+     *  could be found.
+     */
+    private DnsServerAddressStream getNameServersFromCache(String hostname) {
+        int len = hostname.length();
+
+        if (len == 0 || len == 1 && hostname.charAt(0) == '.') {
+            // We never cache for root servers.
+            return null;
+        }
+
+        // We always store in the case with a trailing '.'.
+        if (hostname.charAt(len - 1) != '.') {
+            hostname += ".";
+        }
+
+        // We start from the closed match and then move down.
+        for (;;) {
+            int idx = hostname.indexOf('.');
+            if (idx <= 0) {
+                // We are not interested in handling '.' as we should never serve the root servers from cache.
+                return null;
+            }
+            // Skip '.' as well.
+            hostname = hostname.substring(idx + 1);
+            if (hostname.isEmpty()) {
+                return null;
+            }
+            List<DnsCacheEntry> entries = parent.authoritativeDnsServerCache().get(hostname, additionals);
+            if (entries != null && !entries.isEmpty()) {
+                // Found a match in the cache... Also shuffle them so we not always use the same order for lookup.
+                return DnsServerAddresses.shuffled(new DnsCacheIterable(entries)).stream();
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Avoid instantiation through private constructors from outside of the constructors class. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=pmd%3AAccessorClassGeneration)